### ISSUE TITLE
Introduce ty::WhereClause to align chalk and rustc dyn representation

### ISSUE THREAD

**Issue by csmoe** [rust-lang/compiler-team#433]
# Proposal

As @nikomatsakis wrote in https://github.com/rust-lang/wg-traits/issues/30 , chalk's binder + predicate representation seems better than rustc's [`ExistentialPredicate`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/enum.ExistentialPredicate.html).
https://github.com/rust-lang/rust/pull/85466 renamed `ExistentailPredicate` to `WhereClause`, and created a binder alias:
```rust
pub enum WhereClause {
    ...
}
pub type ExistentialPredicate<'tcx> = Binder<'tcx, WhereClause<'tcx>>;
```
# Mentors or Reviewers

@nikomatsakis 

# Process

The main points of the [Major Change Process][MCP] are as follows:

* [x] File an issue describing the proposal.
* [x] A compiler team member or contributor who is knowledgeable in the area can **second** by writing `@rustbot second`.
    * Finding a "second" suffices for internal changes. If however, you are proposing a new public-facing feature, such as a `-C flag`, then full team check-off is required.
    * Compiler team members can initiate a check-off via `@rfcbot fcp merge` on either the MCP or the PR.
* [ ] Once an MCP is seconded, the Final Comment Period begins. If no objections are raised after 10 days, the MCP is considered **approved**.

You can read [more about Major Change Proposals on forge][MCP].

[MCP]: https://forge.rust-lang.org/compiler/mcp.html

# Comments

**This issue is not meant to be used for technical discussion. There is a Zulip stream for that. Use this issue to leave procedural comments, such as volunteering to review, indicating that you second the proposal (or third, etc), or raising a concern that you would like to be addressed.**
